crates/wasmtime/src/runtime/vm.rs

```diff
@@ -336,6 +336,216 @@ impl ModuleRuntimeInfo {
+    pub fn checked_add(self, bytes: HostAlignedByteCount) -> Result<Self, ByteCountOutOfBounds> {
+        // aligned + aligned = aligned
+        self.0
+            .checked_add(bytes.0)
+            .map(Self)
+            .ok_or(ByteCountOutOfBounds(ByteCountOutOfBoundsKind::Add(
+                self.0, bytes.0,
+            )))
+    }
+
+    /// Subtract two aligned byte counts.
+    ///
+    /// Returns an error if the result underflows.
+    pub fn checked_sub(self, bytes: HostAlignedByteCount) -> Result<Self, ByteCountOutOfBounds> {
+        // aligned - aligned = aligned
+        self.0
+            .checked_sub(bytes.0)
+            .map(Self)
+            .ok_or(ByteCountOutOfBounds(ByteCountOutOfBoundsKind::Sub(
+                self.0, bytes.0,
+            )))
+    }
+
+    /// Multiply an aligned byte count by a scalar value.
+    ///
+    /// Returns an error if the result overflows.
+    pub fn checked_mul(self, n: usize) -> Result<Self, ByteCountOutOfBounds> {
+        // aligned * scalar = aligned
+        self.0
+            .checked_mul(n)
+            .map(Self)
+            .ok_or(ByteCountOutOfBounds(ByteCountOutOfBoundsKind::Mul(
+                self.0, n,
+            )))
+    }
+
+    /// Unchecked multiplication by a scalar value.
+    ///
+    /// ## Safety
+    ///
+    /// The result must not overflow.
+    #[inline]
+    pub unsafe fn unchecked_mul(self, n: usize) -> Self {
+        Self(self.0 * n)
+    }
+}
+
+impl PartialEq<usize> for HostAlignedByteCount {
+    #[inline]
+    fn eq(&self, other: &usize) -> bool {
+        self.0 == *other
+    }
+}
+
+impl PartialEq<HostAlignedByteCount> for usize {
+    #[inline]
+    fn eq(&self, other: &HostAlignedByteCount) -> bool {
+        *self == other.0
+    }
+}
+
+impl fmt::Display for HostAlignedByteCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::LowerHex for HostAlignedByteCount {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::LowerHex::fmt(&self.0, f)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ByteCountNotAligned(usize);
+
+impl fmt::Display for ByteCountNotAligned {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "byte count not page-aligned: {}", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ByteCountNotAligned {}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ByteCountOutOfBounds(ByteCountOutOfBoundsKind);
+
+impl fmt::Display for ByteCountOutOfBounds {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ByteCountOutOfBounds {}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum ByteCountOutOfBoundsKind {
+    RoundUp(usize),
+    U64(u64),
+    Add(usize, usize),
+    Sub(usize, usize),
+    Mul(usize, usize),
```

Comment:
For here I think it's probably ok to discard the precise values that were being operated on as otherwise that makes the errors here pretty large for a low-level type such as this.

Reply:
Hmm good point, done.